client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala

```diff
@@ -111,16 +111,35 @@ class CelebornShuffleReader[K, C](
     val localHostAddress = Utils.localHostName(conf)
     val shuffleKey = Utils.makeShuffleKey(handle.appUniqueId, shuffleId)
     var fileGroups: ReduceFileGroups = null
-    try {
-      // startPartition is irrelevant
-      fileGroups = shuffleClient.updateFileGroup(shuffleId, startPartition)
-    } catch {
-      case ce @ (_: CelebornIOException | _: PartitionUnRetryAbleException) =>
-        // if a task is interrupted, should not report fetch failure
-        // if a task update file group timeout, should not report fetch failure
-        checkAndReportFetchFailureForUpdateFileGroupFailure(shuffleId, ce)
-      case e: Throwable => throw e
-    }
+    var isShuffleStageEnd: Boolean = false
+    var updateFileGroupsRetryTimes = 0
+    do {
+      isShuffleStageEnd =
+        try {
+          shuffleClient.isShuffleStageEnd(shuffleId)
+        } catch {
+          case e: IOException =>
+            logError(s"Failed to check shuffle stage end for $shuffleId, assume not ended", e)
+            false
+        }
+      try {
+        // startPartition is irrelevant
+        fileGroups = shuffleClient.updateFileGroup(shuffleId, startPartition)
+      } catch {
+        case ce: CelebornIOException
+            if ce.getCause.isInstanceOf[TimeoutException] && !isShuffleStageEnd =>
+          updateFileGroupsRetryTimes += 1
+          logError(
+            s"UpdateFileGroup for $shuffleKey timeout due to shuffle stage not ended," +
+              s" retry again, retry times $updateFileGroupsRetryTimes",
+            ce)
+        case ce @ (_: CelebornIOException | _: PartitionUnRetryAbleException) =>
+          // if a task is interrupted, should not report fetch failure
+          // if a task update file group timeout, should not report fetch failure
+          checkAndReportFetchFailureForUpdateFileGroupFailure(shuffleId, ce)
+        case e: Throwable => throw e
+      }
+    } while (true)
```

Comment:
This will indefinitely wait, right ? We should fail after a while.

Reply:
Have updated the code to prevent infinite retry